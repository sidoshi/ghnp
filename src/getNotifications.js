import type { GetNotificationsOptions, GetNotificationsResponse } from './types'

export default async ({
  client,
  url,
  lastModified,
  params = {},
}: GetNotificationsOptions): Promise<GetNotificationsResponse> => {
  // By adding last modified header, we get 304 status if nothing is updated
  const headers =
    lastModified !== undefined ? { 'If-Modified-Since': lastModified } : {}

  try {
    const res = await client.get(url, {
      params,
      headers,
    })

    return {
      notifications: res.data,
      lastModified: res.headers['last-modified'],
      updated: true,
      pollIntervalMs: +res.headers['x-poll-interval'] * 1000 || 0,
    }
  } catch (err) {
    const { response: res } = err

    if (res.status === 304) {
      return {
        notifications: [],
        updated: false,
        lastModified,
        pollIntervalMs: +res.headers['x-poll-interval'] * 1000 || 0,
      }
    }

    throw err
  }
}
