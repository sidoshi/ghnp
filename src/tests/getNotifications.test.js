import getNotifications from '../getNotifications'

const createMockClient = (shouldThrow, status = 200) => ({
  get: jest.fn(() => {
    const res = {
      data: 'NOTIFICATIONS',
      headers: {
        'last-modified': 'LAST_MODIFIED',
        'x-poll-interval': '60',
      },
      status,
    }

    if (shouldThrow) {
      const error = new Error()
      error.response = res
      throw error
    }

    return res
  }),
})

test('Makes a request with proper params and headers', async () => {
  const client = createMockClient()
  const url = 'https://api.github.com'
  const lastModified = new Date().toISOString()
  const params = {
    all: true,
    participating: true,
  }

  const res1 = await getNotifications({ client, url, lastModified, params })
  expect(client.get).toHaveBeenCalledWith(url, {
    headers: { 'If-Modified-Since': lastModified },
    params: { all: true, participating: true },
  })
  expect(res1).toEqual({
    lastModified: 'LAST_MODIFIED',
    notifications: 'NOTIFICATIONS',
    pollIntervalMs: 60000,
    updated: true,
  })

  client.get.mockClear()

  const res2 = await getNotifications({ client, url })
  expect(client.get).toHaveBeenCalledWith(url, { headers: {}, params: {} })
  expect(res2).toEqual({
    lastModified: 'LAST_MODIFIED',
    notifications: 'NOTIFICATIONS',
    pollIntervalMs: 60000,
    updated: true,
  })
})

test('handles 304 Not Changed', async () => {
  const client = createMockClient(true, 304)
  const url = 'https://api.github.com'
  const lastModified = new Date().toISOString()
  const params = {
    all: true,
    participating: true,
  }

  const res = await getNotifications({
    client,
    url,
    lastModified,
    params,
  })

  expect(res).toEqual({
    lastModified,
    notifications: [],
    pollIntervalMs: 60000,
    updated: false,
  })
})

test('Throws on all errors other then 304', async () => {
  const client = createMockClient(true, 500)
  const url = 'https://api.github.com'
  const lastModified = new Date().toISOString()
  const params = {
    all: true,
    participating: true,
  }

  try {
    await getNotifications({
      client,
      url,
      lastModified,
      params,
    })
  } catch (err) {
    expect(err.response).toMatchSnapshot()
  }
})
