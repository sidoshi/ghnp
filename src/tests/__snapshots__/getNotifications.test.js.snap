// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Throws on all errors other then 304 1`] = `
Object {
  "data": "NOTIFICATIONS",
  "headers": Object {
    "last-modified": "LAST_MODIFIED",
    "x-poll-interval": "60",
  },
  "status": 500,
}
`;
